src/library/scala/collection/Factory.scala

```diff
@@ -298,6 +307,16 @@ trait StrictOptimizedSeqFactory[+CC[_]] extends SeqFactory[CC] {
     b.result()
   }
 
+  override def concat[A](xss: Iterable[A]*): CC[A] = {
+    val b = newBuilder[A]
+    val knownSizes = xss.map(_.knownSize)
+    if (knownSizes forall (_ >= 0)) {
+      b.sizeHint(knownSizes.sum)
+    }
```

Comment:
Is this better than the following?

~~~ scala
if (xss.forall(_.knownSize >= 0)) {
  b.sizeHint(xss.view.map(_.knownSize).sum)
}
~~~

Reply:
Ideally, it should be something like this (at most one traversal):
```scala
override def concat[A](xss: Iterable[A]*): CC[A] = {
  def totalKnownSize: Int = {
    val it = xss.iterator
    val size = 0
    // using while loop to avoid non local return
    while (it.hasNext) {
      val knownSize = it.next().knownSize
      if (knownSize < 0) return -1
      size += knownSize
    }
    size
  }

  val knownSize = totalKnownSize
  if (knownSize >= 0)
    b.sizeHint(knownSize)
  ...
}
```
But it might be overkill. I would expect `xss` to have just a couple of elements

Reply:
@allanrenucci Yeah I was also assuming it only had a few elements. I was considering rewriting using your suggestion but wasn't sure if it'd be overkill. You already have to iterate at least a second time after you've provided the size hint.

Reply:
btw, it feels like I should be able to do something like:

```scala
override def concat[A](xss: Iterable[A]*): CC[A] = {
  from(xss.foldLeft(View.empty[A])(new View.Concat(_, _)))
}
```

which seems like it should be nearly fast but much simpler.

Reply:
What’s the advantage compared to the base implementation?

Reply:
> Yeah I was also assuming it only had a few elements

Actually, you could use it as `List.concat(xss: _*)`, and here `xss` might be pretty big. So I would say the optimised implementation is worth the added complexity

Reply:
I don't think there is, because `++` on a `View` will produce a `View.Concat`. But I'm wondering if explicitly giving a size hint is really necessary then, since `View.Concat` should take care of giving us a `knownSize` for the whole thing, which should be utilized by the `IterableFactory` if it makes sense to do so.

So maybe it's not even worth it to add the optimized method?

Reply:
The goal of the override is to use builders, which [have been measured](https://www.scala-lang.org/blog/2018/02/09/collections-performance.html#overhead-of-view-based-implementations) to be ~25% faster than using views.

Reply:
Also, in the current implement, `knownSize` is not constant time for view.
```scala
xss.foldLeft(View.empty[A])(new View.Concat(_, _)).knowSize
```
In the example above, it is linear in the size of `xss`. I'll open an issue